File: src/com/pwootage/fasterflux/blocks/data/BlockIcons.java
package com.pwootage.fasterflux.blocks.data;

import java.util.HashMap;

import net.minecraft.client.renderer.texture.IconRegister;
import net.minecraft.util.Icon;
import net.minecraftforge.common.ForgeDirection;

public abstract class BlockIcons {
	public static String FRONT = "front";
	public static String BACK = "front";
	public static String LEFT = "left";
	public static String RIGHT = "right";
	public static String TOP = "top";
	public static String BOTTOM = "bottom";
	
	private HashMap<String, IconData> icons = new HashMap<>();

	public BlockIcons() {
		init();
	}

	protected void addIcon(String id, IconData icon) {
		icons.put(id, icon);
	}

	/**
	 * Be sure to set all your icons in here!
	 */
	protected abstract void init();

	/**
	 * Register all icons, if not already registered
	 * @param reg IconRegister to register with
	 */
	public void register(IconRegister reg) {
		for (String id : icons.keySet()) {
			IconData icon = icons.get(id);
			icon.register(reg);
		}
	}

	public Icon getIcon(String id) {
		return icons.get(id).get();
	}

	public static final BlockIcons BATTERY_CONTROLLER = new BlockIcons() {
		@Override
		protected void init() {
			addIcon(FRONT, new IconData("fasterflux:batterycontroller"));
		}
	};
	public static final BlockIcons BATTERY_ANODE = new BlockIcons() {
		@Override
		protected void init() {
			addIcon(FRONT, new IconData("fasterflux:anode"));
		}
	};
	public static final BlockIcons BATTERY_CASE = new BlockIcons() {
		@Override
		protected void init() {
			addIcon(FRONT, new IconData("fasterflux:batterycase"));
		}
	};
	public static final BlockIcons UNKNOWN = new BlockIcons() {
		@Override
		protected void init() {
			addIcon(FRONT, new IconData("fasterflux:unknown"));
		}
	};
}

File: src/com/pwootage/fasterflux/blocks/data/BlockIcons.scala
package com.pwootage.fasterflux.blocks.data

import java.util.HashMap
import net.minecraft.client.renderer.texture.IconRegister
import net.minecraft.util.Icon
import net.minecraftforge.common.ForgeDirection
import BlockIcons._
//remove if not needed
import scala.collection.JavaConversions._

object BlockIcons {
  val FRONT: String = "front"
  val BACK: String = "front"
  val LEFT: String = "left"
  val RIGHT: String = "right"
  val TOP: String = "top"
  val BOTTOM: String = "bottom"
  val FLOWING: String = "flowing"
  val STILL: String = "still"

  def register(reg: IconRegister) = {
    for (icon <- ICONS) {
      icon.register(reg)
    }
  }

  private var ICONS = List[BlockIcons]()

  val BATTERY_CONTROLLER = new BlockIcons(Map(
    FRONT -> new IconData("fasterflux:batterycontroller")))

  val BATTERY_ANODE = new BlockIcons(Map(
    FRONT -> new IconData("fasterflux:anode")))

  val BATTERY_CASE = new BlockIcons(Map(
    FRONT -> new IconData("fasterflux:batterycase")))

  val BATTERY_OUT = new BlockIcons(Map(
    FRONT -> new IconData("fasterflux:batteryout")))

  val BATTERY_IN = new BlockIcons(Map(
    FRONT -> new IconData("fasterflux:batteryin")))

  val LIQUID_NETHER = new BlockIcons(Map(
    FLOWING -> new IconData("fasterflux:liquidnetherflow"),
    STILL -> new IconData("fasterflux:liquidnetherstill")))

  val UNKNOWN = new BlockIcons(Map(
    FRONT -> new IconData("fasterflux:unknown")))

  class BlockIcons(icons: Map[String, IconData]) {
    ICONS = ICONS :+ this

    def register(reg: IconRegister) {
      for ((id, icon) <- icons) {
        icon.register(reg)
      }
    }

    def getIcon(id: String): Icon = getIconData(id) match {
      case null => null
      case icon => icon get
    }

    def getIconData(id: String): IconData = icons get id match {
      case None => null
      case icon => icon getOrElse null
    }
  }
}

